OpenRA.Mods.Common/Traits/Buildings/DockClientManager.cs

```diff
@@ -0,0 +1,298 @@
+#region Copyright & License Information
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Mods.Common.Orders;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Manages DockClients on the actor.")]
+	public class DockClientManagerInfo : ConditionalTraitInfo
+	{
+		[Desc("Go to DockTypes on deploy.")]
+		public readonly BitSet<DockType> ReturnToBaseDockType;
+
+		[Desc("How long (in ticks) to wait until (re-)checking for a nearby available DockHost.")]
+		public readonly int SearchForDockDelay = 125;
+
+		[Desc("The pathfinding cost penalty applied for each dock client waiting to unload at a DockHost.")]
+		public readonly int OccupancyCostModifier = 12;
+
+		[CursorReference]
+		[Desc("Cursor to display when able to dock at target actor.")]
+		public readonly string EnterCursor = "enter";
+
+		[CursorReference]
+		[Desc("Cursor to display when unable to dock at target actor.")]
+		public readonly string EnterBlockedCursor = "enter-blocked";
+
+		[VoiceReference]
+		[Desc("Voice.")]
+		public readonly string Voice = "Action";
+
+		[Desc("Color to use for the target line of docking orders.")]
+		public readonly Color DockLineColor = Color.Green;
+
+		public override object Create(ActorInitializer init) { return new DockClientManager(init.Self, this); }
+	}
+
+	public class DockClientManager : ConditionalTrait<DockClientManagerInfo>, IResolveOrder, IOrderVoice, IIssueOrder, INotifyActorDisposing
+	{
+		readonly Actor self;
+		protected IDockClient[] dockClients;
+		public bool IsAliveAndInWorld => !self.IsDead && self.IsInWorld;
+		public Color DockLineColor => Info.DockLineColor;
+		public int OccupancyCostModifier => Info.OccupancyCostModifier;
+
+		public DockClientManager(Actor self, DockClientManagerInfo info)
+			: base(info)
+		{
+			this.self = self;
+		}
+
+		protected override void Created(Actor self)
+		{
+			base.Created(self);
+			dockClients = self.TraitsImplementing<IDockClient>().ToArray();
+		}
+
+		public Actor ReservedHostActor { get; protected set; }
+		public DockHost ReservedHost { get; protected set; }
+
+		DockHost lastReservedDockHost = null;
+		public DockHost LastReservedHost
```

Comment:
A one line comment explaining what the "LastReservedHost" is would be good, as this is not immediately clear.

Reply:
The variable name is literal. I'm not sure how clearer can you get

Reply:
Ok, I think you're right, though the problem I am having is understanding *why* the last reserved host is used, since being recently used doesn't immediately seem to be meaningful. `GetSearchFromProcLocation()` is where I didn't understand why you used this variable, so perhaps documenting that function, as well as other places where the variable is used in an unusual way (i.e. not for `UnreserveHost()`, since that is obvious).

Reply:
I think it's explained fairly well at places the value is used

Reply:
Added a bunch more documentation

Reply:
Ok I think it's fine now, it was the multiple layers of logic that confused me. The documentation helps a lot.